hw/xbox/xid-sdl.c

```diff
@@ -194,36 +195,40 @@ static const XIDDesc desc_xid_xbox_gamepad = {
 
 #define BUTTON_MASK(button) (1 << ((button) - GAMEPAD_DPAD_UP))
 
-static void update_output(USBXIDState *s) {
-#ifdef FORCE_FEEDBACK
-    if (s->sdl_haptic == NULL) { return; }
-
-    SDL_HapticLeftRight effect = {
-        .type = SDL_HAPTIC_LEFTRIGHT,
-        .length = SDL_HAPTIC_INFINITY,
-        /* FIXME: Might be left/right inverted */
-        .large_magnitude = s->out_state.right_actuator_strength,
-        .small_magnitude = s->out_state.left_actuator_strength
-    };
-
-    if (s->sdl_haptic_effect_id == -1) {
-        int effect_id = SDL_HapticNewEffect(s->sdl_haptic,
-                                            (SDL_HapticEffect*)&effect);
-        if (effect_id == -1) {
-            fprintf(stderr, "Failed to upload haptic effect!\n");
-            SDL_HapticClose(s->sdl_haptic);
-            s->sdl_haptic = NULL;
-            return;
-        }
-        SDL_HapticRunEffect(s->sdl_haptic, effect_id, 1);
+static SDL_HapticEffect haptic_effect_init(USBXIDState *s, float large_magnitude_percent, float small_magnitude_percent) {
+    if (large_magnitude_percent > 1) {
+        large_magnitude_percent = 1;
+    }
 
-        s->sdl_haptic_effect_id = effect_id;
+    if (small_magnitude_percent > 1) {
+        small_magnitude_percent = 1;
+    }
 
+    SDL_HapticEffect effect;
+    if (s->sdl_haptic_effect_type == SDL_HAPTIC_LEFTRIGHT) {
+        effect.leftright = (SDL_HapticLeftRight){
+            .type = SDL_HAPTIC_LEFTRIGHT,
+            .length = SDL_HAPTIC_INFINITY,
+            .large_magnitude = floor(large_magnitude_percent * SDL_MAX_HAPTIC_VALUE),
+            .small_magnitude = floor(small_magnitude_percent * SDL_MAX_HAPTIC_VALUE),
+        };
     } else {
-        SDL_HapticUpdateEffect(s->sdl_haptic, s->sdl_haptic_effect_id,
-                               (SDL_HapticEffect*)&effect);
+        fprintf(stderr, "Unknown haptic effect type\n");
     }
-#endif
+
+    return effect;
+}
+
+static void update_output(USBXIDState *s) {
+    if (s->sdl_haptic == NULL) { return; }
+
+    SDL_HapticEffect effect = haptic_effect_init(
+        s,
+        s->out_state.right_actuator_strength / (float)XID_MAX_HAPTIC_VALUE,
+        s->out_state.left_actuator_strength / (float)XID_MAX_HAPTIC_VALUE
+    );
```

Comment:
nit: Why bother with floats?

These are obviously integers using their full range; so to go from XID to SDL values, we just have to do `actuator_strength >> 1`.

*(We do the same for input, but it might actually change there because input must be more flexible - for output I don't see this issue)*

Reply:
Floats felt more generic to me, `actuator_strength >> 1` relies on xbox xid out_state specifics. But I see your point on output not needing to be flexible and I like the implementation much more without the floats, so I'll follow your advice here.